### PR TITLE
Vpnaas: Create VPN service

### PR DESCRIPTION
**Pull request by simonre**
For #723

Links to the line numbers/files in the OpenStack source code that support the
code in this PR:

* https://github.com/openstack/neutron-vpnaas/blob/058469e1b99b647537a5228c6a384d93df5484df/neutron_vpnaas/db/vpn/vpn_db.py#L473
* https://github.com/openstack/neutron-vpnaas/blob/058469e1b99b647537a5228c6a384d93df5484df/neutron_vpnaas/db/vpn/vpn_db.py#L459

API:
https://developer.openstack.org/api-ref/network/v2/#create-vpn-service

### CODE REVIEW

**Comment by simonre** [comment]
@jtopjian This is ready for review

File: openstack/networking/v2/extensions/vpnaas/services/requests.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import "github.com/gophercloud/gophercloud"
+
+// CreateOptsBuilder allows extensions to add additional parameters to the
+// Create request.
+type CreateOptsBuilder interface {
+	ToVPNServiceCreateMap() (map[string]interface{}, error)
+}
+
+// CreateOpts contains all the values needed to create a new VPN service
+type CreateOpts struct {
```
Comment (jtopjian) [inline openstack/networking/v2/extensions/vpnaas/services/requests.go]: doc nit: all fields should have a descriptive comment even if it's something simple like:

```
Name is the human-readable name of the VPN Service.
```

File: openstack/networking/v2/extensions/vpnaas/services/requests.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+	// TenantID specifies a tenant to own the VPN service. The caller must have
+	// an admin role in order to set this. Otherwise, this field is left unset
+	// and the caller will be the owner.
+	TenantID     string `json:"tenant_id,omitempty"`
+	SubnetID     string `json:"subnet_id,omitempty"`
+	RouterID     string `json:"router_id" required:"true"`
+	Description  string `json:"description,omitempty"`
+	AdminStateUp *bool  `json:"admin_state_up"`
+	ProjectID    string `json:"project_id"`
```
Comment (jtopjian) [inline openstack/networking/v2/extensions/vpnaas/services/requests.go]: I don't see `project_id` defined anywhere [here](https://github.com/openstack/neutron-vpnaas/blob/058469e1b99b647537a5228c6a384d93df5484df/neutron_vpnaas/db/vpn/vpn_db.py#L473), so maybe it's an internal neutron field? What happens if `TenantID` and `ProjectID` are different UUIDs?

File: openstack/networking/v2/extensions/vpnaas/services/requests.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
+package services
+
+import "github.com/gophercloud/gophercloud"
+
+// CreateOptsBuilder allows extensions to add additional parameters to the
+// Create request.
+type CreateOptsBuilder interface {
+	ToVPNServiceCreateMap() (map[string]interface{}, error)
```
Comment (jtopjian) [inline openstack/networking/v2/extensions/vpnaas/services/requests.go]: This should be `ToServiceCreateMap`.

File: openstack/networking/v2/extensions/vpnaas/services/requests.go [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+	RouterID     string `json:"router_id" required:"true"`
+	Description  string `json:"description,omitempty"`
+	AdminStateUp *bool  `json:"admin_state_up"`
+	ProjectID    string `json:"project_id"`
+	Name         string `json:"name,omitempty"`
+	FlavorID     string `json:"flavor_id,omitempty"`
+}
+
+// ToVPNServiceCreateMap casts a CreateOpts struct to a map.
+func (opts CreateOpts) ToVPNServiceCreateMap() (map[string]interface{}, error) {
```
Comment (jtopjian) [inline openstack/networking/v2/extensions/vpnaas/services/requests.go]: `ToServiceCreateMap`

File: openstack/networking/v2/extensions/vpnaas/services/results.go [hunk 5]
```diff
@@ -0,0 +1,40 @@
+package services
+
+import (
+	"github.com/gophercloud/gophercloud"
+)
+
+// Service is a VPN Service
+type Service struct {
```
Comment (jtopjian) [inline openstack/networking/v2/extensions/vpnaas/services/results.go]: doc nit: these fields should have comments to describe what they are.

File: openstack/networking/v2/extensions/vpnaas/services/results.go [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+	TenantID     string `json:"tenant_id"`
+	SubnetID     string `json:"subnet_id,omitempty"`
```
Comment (jtopjian) [inline openstack/networking/v2/extensions/vpnaas/services/results.go]: `omitempty` isn't valid for results. it can safely be removed.

**Comment by jtopjian** [comment]
@simonre Thank you for the updates. Travis is failing due to some `gofmt` issues. These can be fixed by manually running `gofmt` on the changed files or configuring your IDE/editor to run `gofmt` upon save.

The OpenLab failures were a temporary issue with the CI system. I wouldn't worry about those.

**Comment by jtopjian** [comment]
@simonre Thank you again for the updates.

This looks good to me. Unfortunately I can't test this myself since I don't have access to VPNaaS, but from reviewing the Neutron code, this looks correct. In addition, the acceptance test looks correct, too.

Can you verify that this code works? Meaning: have you tested it yourself to create an actual VPN service? If it works for you, I'll go ahead and merge this.

The inline comments/docs could use some formatting, but I don't consider doc updates to be blocking unless there are other changes required. If you plan on continuing to work on the other parts of the VPNaaS, then maybe fix them there. You can use [this](https://github.com/gophercloud/gophercloud/blob/master/openstack/networking/v2/extensions/subnetpools/requests.go) file for reference (or really any of the `requests.go` and `results.go` files). Note the formatting of the comments for the fields, particularly the space between `//` and the comment as well as a new line to separate the fields.

The reason we want to make sure formatting is correct is because the inline comments turn into the official godoc reference documentation.

**Comment by simonre** [comment]
@jtopjian Thanks for the quick feedback. I just updated the doc.go sample code to code that I can verify worked for me to create an actual VPN service. I also updated the inline formatting in requests.go, is that what you had in mind?

**Comment by jtopjian** [comment]
@simonre 

>  I also updated the inline formatting in requests.go, is that what you had in mind?

Yep - That's exactly what I had in mind. Though the same format should be applied to the `Service` result struct, too. Let me know if you want to fix that now or in another PR.

> I can verify worked for me to create an actual VPN service

Perfect - thank you 😄

**Comment by simonre** [comment]
I'll fix it now. Just wanted to make sure I'm actually doing the right thing before doing it for all files.

**Comment by jtopjian** [comment]
@simonre I apologize - I missed one thing in the unit tests:

There should be a comparison to the struct that is returned from the Create / Extract call:

```go
expected := services.Service{
  RouterID: "66e3b16c-8ce5-40fb-bb49-ab6d8dc3f2aa",
  ...
}

actual, err := services.Create(fake.ServiceClient(), options).Extract()
th.AssertNoErr(t, err)
th.AssertDeepEquals(t, expected, actual)
```

This same pattern will need to be in the other unit tests (where applicable), so best to get in the habit now :)

**Comment by simonre** [comment]
@jtopjian 
`th.AssertDeepEquals(t, expected, actual)`
doesn't seem to work because of the pointer to AdminStateUp. I added individual comparisons like the ones [here](https://github.com/gophercloud/gophercloud/blob/cf2bbaa52ab69d063466a3c46a5918d05d752625/openstack/networking/v2/extensions/fwaas/firewalls/testing/requests_test.go#L224) instead which should do the same.

**Comment by jtopjian** [comment]
@simonre We definitely want to ensure a struct can be compared successfully, and not just individual fields.

Using "networks" as an example, the `AdminStateUp` field for the result struct is `bool` rather than `*bool`:

https://github.com/gophercloud/gophercloud/blob/master/openstack/networking/v2/networks/results.go#L57

However, the CreateOpts `AdminStateUp` should probably remain `*bool` as this is the standard used for the networking services. For example:

https://github.com/gophercloud/gophercloud/blob/master/openstack/networking/v2/networks/requests.go#L69

Unless there's a good reason to keep the result field as `*bool`, let's change it to `bool`. The only reason I can think of using `*bool` is if there are three distinct states for `AdminStateUp` (true/false/nil). Given the prevalence of `AdminStateUp` already in Gophercloud, I can't see VPNaaS doing something different.

**Comment by simonre** [comment]
@jtopjian You're right. I changed it again. I also did the same in the ipsecpolicy branch.

**Comment by jtopjian** [comment]
@simonre This looks good to me. Thank you very much for your help and patience.

**Pull request merged**